java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -32,32 +32,38 @@ class ApacheHttpRequest extends RefType {
 class URLConstructor extends ClassInstanceExpr {
   URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
 
-  // URLs constructed with the string constructor `URL(String spec)`
-  Expr specArg() {
+  predicate hasHttpStringArg() {
     this.getConstructor().getParameter(0).getType() instanceof TypeString and
     (
+      // URLs constructed with the string constructor `URL(String spec)`
       this.getConstructor().getNumberOfParameters() = 1 and
-      result = this.getArgument(0) // First argument contains the whole spec.
-    )
-  }
-
-  // URLs constructed with any of the three string constructors below:
-  // `URL(String protocol, String host, int port, String file)`,
-  // `URL(String protocol, String host, int port, String file, URLStreamHandler handler)`,
-  // `URL(String protocol, String host, String file)`
-  Expr protocolArg() {
-    this.getConstructor().getParameter(0).getType() instanceof TypeString and
-    (
+      this.getArgument(0) instanceof HttpString // First argument contains the whole spec.
+      or
+      // URLs constructed with any of the three string constructors below:
+      // `URL(String protocol, String host, int port, String file)`,
+      // `URL(String protocol, String host, int port, String file, URLStreamHandler handler)`,
+      // `URL(String protocol, String host, String file)`
       this.getConstructor().getNumberOfParameters() > 1 and
-      result = this.getArgument(0) // First argument contains the protocol part.
+      concatHttpString(getArgument(0), this.getArgument(1)) // First argument contains the protocol part and the second argument contains the host part.
     )
   }
+}
 
-  Expr hostArg() {
-    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+/**
+ * Class of Java URI constructor.
+ */
+class URIConstructor extends ClassInstanceExpr {
+  URIConstructor() { this.getConstructor().getDeclaringType().hasQualifiedName("java.net", "URI") }
+
+  predicate hasHttpStringArg() {
     (
-      this.getConstructor().getNumberOfParameters() > 1 and
-      result = this.getArgument(1) // Second argument contains the host part.
+      this.getNumArgument() = 1 // `URI(String str)`
```

Comment:
Is this missing the check for the argument?
```suggestion
      this.getNumArgument() = 1 and
      this.getArgument(0) instanceof HttpString // `URI(String str)`
```

Reply:
Check for the argument is done on line 176 in the `createURI` predicate, which is as follows:
```
cc.getArgument(0) = node1.asExpr() and
```

Reply:
But that only makes sure that there is taint flow, doesn't it?
I think here this would be incorrectly considered a source:
https://github.com/github/codeql/blob/b65a03330243a4f7565487f2320b305bf16bd03b/java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql#L229-L232

Reply:
Sorry @Marcono1234 I don't quite get your point. The reason I created the `URIConstructor` source above is that it requires to check inputs from two arguments together, which is not of the `Expr` type `HttpString`.

The current implementation works although it may not be optimal.

Could you elaborate your concerns and suggestion on possible improvements?

Thanks,
@luchua-bc

Reply:
The predicate `URIConstructor.hasHttpStringArg()` matches in line 60 if the constructor call has one argument, not verifying that it is actually a `HttpString` and in line 229 this predicate is used to determine whether an expression is a taint source.

I don't have a test setup for this query locally, but I think it would currently cause a false positive for the following. Could you please check it?
```java
// Notice how this is using a different scheme
URI uri = new URI("any://does-not-matter");
HttpRequestBase post = new HttpPost(uri);

String authString = username + ":" + password;
byte[] authEncBytes = Base64.getEncoder().encode(authString.getBytes());
String authStringEnc = new String(authEncBytes);

post.addHeader("Authorization", "Basic " + authStringEnc);
```

Reply:
Thanks @Marcono1234. You are right, I must have missed this test scenario after revamping the code. I've changed the code as per your suggestion.